File: backend/app/core/premium.py
"""Premium presentation policy. Existing free customizations are grandfathered.

The first premium save keeps an immutable free presentation snapshot. Expiration
changes the public projection, never the stored customization. All profile write
paths (including templates) pass through the shared persistence boundary.
"""
from copy import deepcopy
import re
from fastapi import HTTPException

SHARED_SETTINGS = ("layout", "pageEnter", "profileFont", "profileFontScope", "entryScreen", "entryText", "clickSound", "bioTypewriter", "bioTypeMs", "bioDeleteMs", "bioPauseMs", "cardTilt", "showViews", "showAvatar", "showSocials", "borderColor", "borderWidth", "profileRadius", "ogTitle", "ogDescription", "ogOverlayAvatar", "ogOverlayName", "ogOverlayAddress")
PREMIUM_ASSETS = ("customFont", "clickSound", "entryIcon", "ogImage", "favicon")
DEFAULTS = {"version": 1, "cursorEffect": "None", "cursorColor": "#ffffff", "clickPreset": "None", "entrySubtitle": "", "typewriterTexts": [], "hero": "Classic", "borderType": "Static", "borderOpacity": 100, "borderEnabled": True}


async def has_premium(user_id: str) -> bool:
    from app.db import admin_db
    if not admin_db.has_pool():
        return False
    return await admin_db.has_active_premium(user_id)


def normalize_premium(raw):
    from app.core.profile_sanitize import css_hex_color, _plain_text, _clamp_int
    if not isinstance(raw, dict):
        return None
    result = dict(DEFAULTS)
    for key, allowed in {"cursorEffect": ("None", "Cursor Cat", "Snowflakes", "Ghost Cursor", "Following Dot", "Bubbles"), "clickPreset": ("None", "Crisp Click", "Pixel Click", "Bass Tick", "Mouse Click", "Custom"), "hero": ("Classic", "Centered"), "borderType": ("Static", "Dashed", "Shimmer", "Pulse")}.items():
        result[key] = raw.get(key) if raw.get(key) in allowed else DEFAULTS[key]
    result["cursorColor"] = css_hex_color(raw.get("cursorColor"), "#ffffff")
    result["entrySubtitle"] = _plain_text(raw.get("entrySubtitle"), 160)
    result["borderOpacity"] = _clamp_int(raw.get("borderOpacity"), 100, 0, 100)
    result["borderEnabled"] = raw.get("borderEnabled") is not False
    result["typewriterTexts"] = [_plain_text(s, 200) for s in raw.get("typewriterTexts", [])[:12] if isinstance(s, str) and s.strip()] if isinstance(raw.get("typewriterTexts"), list) else []
    colors = raw.get("effectColors")
    result["effectColors"] = {key: value.lower() if re.fullmatch(r"#[0-9a-fA-F]{6}", value) else default for key, default in {"Fireflies": "#fcd271", "Snowflakes": "#ffffff", "Snow": "#ffffff", "Sakura": "#ffb7c5"}.items() if isinstance(colors, dict) and isinstance((value := colors.get(key)), str)}
    return result


def _presentation(config):
    return {"settings": {key: config.get("settings", {}).get(key) for key in SHARED_SETTINGS}, "assets": {key: config.get("assets", {}).get(key) for key in PREMIUM_ASSETS}, "sections": deepcopy(config.get("sections", []))}


def protect_write(config, previous, entitled):
    from app.core.profile_sanitize import sanitize_profile_config
    incoming = sanitize_profile_config(config)
    previous = previous or {}
    old = sanitize_profile_config(previous)
    advanced = incoming["settings"].get("premium")
    base = previous.get("_premium_base")
    changed = advanced != old["settings"].get("premium")
    changed |= incoming["settings"].get("layout") in ("Default", "Portfolio") and incoming["settings"].get("layout") != old["settings"].get("layout")
    changed |= any(s.get("leftCard") or s.get("rightCard") or s.get("subtitle") or (s.get("type") == "about" and s.get("tags")) or s.get("type") == "integration" for s in incoming.get("sections", [])) and incoming.get("sections") != old.get("sections")
    changed |= (incoming.get("assets", {}).get("entryIcon") or {}).get("url") != (old.get("assets", {}).get("entryIcon") or {}).get("url")
    if base:
        changed |= _presentation(incoming) != _presentation(old)
    if changed and not entitled:
        raise HTTPException(403, "An active Premium entitlement is required to change premium settings. Your saved settings have been kept.")
    incoming.pop("_premium_base", None)
    if base:
        incoming["_premium_base"] = deepcopy(base)
    elif entitled and advanced:
        incoming["_premium_base"] = _presentation(old)
    return incoming


def public_projection(config, entitled):
    result = deepcopy(config)
    base = result.pop("_premium_base", None)
    if not entitled:
        result.setdefault("settings", {}).pop("premium", None)
        if base:
            for group in ("settings", "assets"):
                for key, value in base.get(group, {}).items():
                    if value is None:
                        result.setdefault(group, {}).pop(key, None)
                    else:
                        result.setdefault(group, {})[key] = value
            result["sections"] = base.get("sections", [])
        elif result["settings"].get("layout") in ("Default", "Portfolio"):
            result["settings"]["layout"] = "Modern"
    return result

File: frontend/lib/server/premium.ts
import "server-only";
import { one } from "./postgres";

const SHARED_SETTINGS = [
  "layout", "pageEnter", "profileFont", "profileFontScope", "entryScreen",
  "entryText", "clickSound", "bioTypewriter", "bioTypeMs", "bioDeleteMs",
  "bioPauseMs", "cardTilt", "showViews", "showAvatar", "showSocials",
  "borderColor", "borderWidth", "profileRadius", "ogTitle", "ogDescription",
  "ogOverlayAvatar", "ogOverlayName", "ogOverlayAddress"
] as const;

const PREMIUM_ASSETS = [
  "customFont", "clickSound", "entryIcon", "ogImage", "favicon"
] as const;

export async function hasActivePremium(userId: string): Promise<boolean> {
  try {
    const row = await one<{ active: boolean }>(`
      SELECT EXISTS (
        SELECT 1 FROM premium_entitlements
        WHERE user_id = $1 AND active = TRUE AND (expires_at IS NULL OR expires_at > NOW())
      ) AS active
    `, [userId]);
    return Boolean(row?.active);
  } catch {
    return false;
  }
}

function presentationSnapshot(config: Record<string, unknown>): Record<string, unknown> {
  const settings = (config.settings && typeof config.settings === "object" ? config.settings : {}) as Record<string, unknown>;
  const assets = (config.assets && typeof config.assets === "object" ? config.assets : {}) as Record<string, unknown>;
  const sections = Array.isArray(config.sections) ? structuredClone(config.sections) : [];

  const snapSettings: Record<string, unknown> = {};
  for (const key of SHARED_SETTINGS) {
    snapSettings[key] = settings[key];
  }

  const snapAssets: Record<string, unknown> = {};
  for (const key of PREMIUM_ASSETS) {
    snapAssets[key] = assets[key];
  }

  return {
    settings: snapSettings,
    assets: snapAssets,
    sections,
  };
}

export function protectWrite(
  incoming: Record<string, unknown>,
  previous: Record<string, unknown> | null,
  entitled: boolean
): Record<string, unknown> {
  const prev = previous || {};
  const inSettings = (incoming.settings && typeof incoming.settings === "object" ? incoming.settings : {}) as Record<string, unknown>;
  const oldSettings = (prev.settings && typeof prev.settings === "object" ? prev.settings : {}) as Record<string, unknown>;
  const inAssets = (incoming.assets && typeof incoming.assets === "object" ? incoming.assets : {}) as Record<string, unknown>;
  const oldAssets = (prev.assets && typeof prev.assets === "object" ? prev.assets : {}) as Record<string, unknown>;
  const inSections = Array.isArray(incoming.sections) ? incoming.sections : [];
  const oldSections = Array.isArray(prev.sections) ? prev.sections : [];

  const advanced = inSettings.premium;
  const base = prev._premium_base as Record<string, unknown> | undefined;

  let changed = JSON.stringify(advanced) !== JSON.stringify(oldSettings.premium);

  if (["Default", "Portfolio"].includes(String(inSettings.layout || "")) && inSettings.layout !== oldSettings.layout) {
    changed = true;
  }

  const hasAdvancedSections = inSections.some((s: any) =>
    s && typeof s === "object" && (s.leftCard || s.rightCard || s.subtitle || (s.type === "about" && s.tags) || s.type === "integration")
  );
  if (hasAdvancedSections && JSON.stringify(inSections) !== JSON.stringify(oldSections)) {
    changed = true;
  }

  const inEntryIcon = (inAssets.entryIcon as Record<string, unknown> | undefined)?.url;
  const oldEntryIcon = (oldAssets.entryIcon as Record<string, unknown> | undefined)?.url;
  if (inEntryIcon !== oldEntryIcon) {
    changed = true;
  }

  if (base) {
    const currSnap = presentationSnapshot(incoming);
    const oldSnap = presentationSnapshot(prev);
    if (JSON.stringify(currSnap) !== JSON.stringify(oldSnap)) {
      changed = true;
    }
  }

  if (changed && !entitled) {
    throw new Error("An active Premium entitlement is required to change premium settings. Your saved settings have been kept.");
  }

  const result = { ...incoming };
  delete result._premium_base;

  if (base) {
    result._premium_base = structuredClone(base);
  } else if (entitled && advanced) {
    result._premium_base = presentationSnapshot(prev);
  }

  return result;
}

export function publicProjection(
  config: Record<string, unknown>,
  entitled: boolean
): Record<string, unknown> {
  const result = structuredClone(config);
  const base = result._premium_base as Record<string, unknown> | undefined;
  delete result._premium_base;

  if (!entitled) {
    if (result.settings && typeof result.settings === "object") {
      delete (result.settings as Record<string, unknown>).premium;
    }

    if (base) {
      for (const group of ["settings", "assets"] as const) {
        const groupObj = (base[group] && typeof base[group] === "object" ? base[group] : {}) as Record<string, unknown>;
        if (!result[group] || typeof result[group] !== "object") {
          result[group] = {};
        }
        const targetGroup = result[group] as Record<string, unknown>;
        for (const [key, value] of Object.entries(groupObj)) {
          if (value === null || value === undefined) {
            delete targetGroup[key];
          } else {
            targetGroup[key] = value;
          }
        }
      }
      result.sections = Array.isArray(base.sections) ? structuredClone(base.sections) : [];
    } else {
      const settings = (result.settings && typeof result.settings === "object" ? result.settings : {}) as Record<string, unknown>;
      if (["Default", "Portfolio"].includes(String(settings.layout || ""))) {
        settings.layout = "Modern";
      }
    }
  }

  return result;
}
